### ISSUE TITLE
CUDA: Invalid global accesses with for loops and indexing

### ISSUE THREAD

**Issue by gmarkall** [numba/numba#5576]
The following (reduced from `test_issue_5073`):

```python
from numba import cuda
try:
    from numba.np import numpy_support as nps
except ImportError:
    from numba import numpy_support as nps

import numpy as np

arr = np.arange(1024)
nthreads = 16
dt = nps.from_dtype(arr.dtype)
nshared = nthreads * arr.dtype.itemsize

@cuda.jit
def sm_slice_copy(y, chunksize):
    dynsmem = cuda.shared.array(0, dtype=dt)

    bx = cuda.blockIdx.x

    for j in range(chunksize):
        y[bx + j] = dynsmem[j]


d_result = cuda.device_array_like(arr)
sm_slice_copy[1, 1, 0, nshared](d_result, 8)
host_result = d_result.copy_to_host()
np.testing.assert_array_equal(arr, host_result)
```

when run with:

```
cuda-memcheck python repro.py
```

produces:

```
========= CUDA-MEMCHECK
========= Invalid __global__ write of size 8
=========     at 0x000008b0 in cudapy::__main__::sm_slice_copy$241(Array<__int64, int=1, C, mutable, aligned>, __int64)
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x7f2369600008 is out of bounds
=========     Device Frame:cudapy::__main__::sm_slice_copy$241(Array<__int64, int=1, C, mutable, aligned>, __int64) (cudapy::__main__::sm_slice_copy$241(Array<__int64, int=1, C, mutable, aligned>, __int64) : 0x8b0)
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so (cuLaunchKernel + 0xb7c) [0x836206]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/../../libffi.so.6 (ffi_call_unix64 + 0x4c) [0x5ec0]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/../../libffi.so.6 (ffi_call + 0x22d) [0x587d]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so (_ctypes_callproc + 0x2ce) [0x1201e]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so [0x12a54]
=========     Host Frame:python (PyObject_Call + 0x6e) [0x127fde]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1e9d) [0x1cd54d]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0xab8) [0x115b48]
=========     Host Frame:python (_PyFunction_FastCallKeywords + 0x387) [0x1692b7]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x4b39) [0x1d01e9]
=========     Host Frame:python (_PyFunction_FastCallKeywords + 0xfb) [0x16902b]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x416) [0x1cbac6]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x115389]
=========     Host Frame:python (_PyFunction_FastCallDict + 0x1d5) [0x1164c5]
=========     Host Frame:python (_PyObject_Call_Prepend + 0xde) [0x135aee]
=========     Host Frame:python [0x17d27a]
=========     Host Frame:python (PyObject_Call + 0x6e) [0x127fde]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1e9d) [0x1cd54d]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0xab8) [0x115b48]
=========     Host Frame:python (_PyFunction_FastCallKeywords + 0x387) [0x1692b7]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x14d4) [0x1ccb84]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x115389]
=========     Host Frame:python (_PyFunction_FastCallDict + 0x1d5) [0x1164c5]
=========     Host Frame:python (_PyObject_Call_Prepend + 0x63) [0x135a73]
=========     Host Frame:python [0x17d27a]
=========     Host Frame:python (PyObject_Call + 0x6e) [0x127fde]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1e9d) [0x1cd54d]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x115389]
=========     Host Frame:python (_PyFunction_FastCallDict + 0x1d5) [0x1164c5]
=========     Host Frame:python (_PyObject_Call_Prepend + 0x63) [0x135a73]
=========     Host Frame:python [0x17d27a]
=========     Host Frame:python (_PyObject_FastCallKeywords + 0x48b) [0x17e2db]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x4a96) [0x1d0146]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x115389]
=========     Host Frame:python (PyEval_EvalCodeEx + 0x44) [0x1162b4]
=========     Host Frame:python (PyEval_EvalCode + 0x1c) [0x1162dc]
=========     Host Frame:python [0x22c664]
=========     Host Frame:python (PyRun_FileExFlags + 0xa1) [0x236a91]
=========     Host Frame:python (PyRun_SimpleFileExFlags + 0x1c3) [0x236c83]
=========     Host Frame:python [0x237db5]
=========     Host Frame:python (_Py_UnixMain + 0x3c) [0x237edc]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xe7) [0x21b97]
=========     Host Frame:python [0x1db3e0]
=========
========= Program hit CUDA_ERROR_LAUNCH_FAILED (error 719) due to "unspecified launch failure" on CUDA API call to cuMemcpyDtoH_v2. 
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so (cuMemcpyDtoH_v2 + 0x6dc) [0x816729]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/../../libffi.so.6 (ffi_call_unix64 + 0x4c) [0x5ec0]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/../../libffi.so.6 (ffi_call + 0x22d) [0x587d]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so (_ctypes_callproc + 0x2ce) [0x1201e]
=========     Host Frame:/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so [0x12a54]
=========     Host Frame:python (PyObject_Call + 0x6e) [0x127fde]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1e9d) [0x1cd54d]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0xab8) [0x115b48]
=========     Host Frame:python (_PyFunction_FastCallDict + 0x1d5) [0x1164c5]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1e9d) [0x1cd54d]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x115389]
=========     Host Frame:python (_PyFunction_FastCallKeywords + 0x387) [0x1692b7]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x14d4) [0x1ccb84]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x115389]
=========     Host Frame:python (_PyFunction_FastCallDict + 0x1d5) [0x1164c5]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1e9d) [0x1cd54d]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x5da) [0x11566a]
=========     Host Frame:python (_PyFunction_FastCallKeywords + 0x387) [0x1692b7]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x690) [0x1cbd40]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x115389]
=========     Host Frame:python (PyEval_EvalCodeEx + 0x44) [0x1162b4]
=========     Host Frame:python (PyEval_EvalCode + 0x1c) [0x1162dc]
=========     Host Frame:python [0x22c664]
=========     Host Frame:python (PyRun_FileExFlags + 0xa1) [0x236a91]
=========     Host Frame:python (PyRun_SimpleFileExFlags + 0x1c3) [0x236c83]
=========     Host Frame:python [0x237db5]
=========     Host Frame:python (_Py_UnixMain + 0x3c) [0x237edc]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xe7) [0x21b97]
=========     Host Frame:python [0x1db3e0]
=========
Traceback (most recent call last):
  File "repro.py", line 26, in <module>
    host_result = d_result.copy_to_host()
  File "/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 225, in _require_cuda_context
    return fn(*args, **kws)
  File "/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/site-packages/numba/cuda/cudadrv/devicearray.py", line 268, in copy_to_host
    _driver.device_to_host(hostary, self, self.alloc_size, stream=stream)
  File "/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 1918, in device_to_host
    fn(host_pointer(dst), device_pointer(src), size, *varargs)
  File "/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/gmarkall/miniconda3/envs/numba048/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [719] Call to cuMemcpyDtoH results in CUDA_ERROR_LAUNCH_FAILED
========= ERROR SUMMARY: 2 errors
```

This happens with both Numba 0.48 and current master. However, I can only reproduce it on an RTX 8000 so far.

For some reason the write to `y[bx + j]` is quite a long way before the start of memory for the array - `0xffffffb8` bytes before it - it looks a bit like some sort of signed vs unsigned arithmetic error in the computation of the index.

**Comment by uchytilc** [comment]
I just submitted a bug reported but closed it because I saw this and the two might be related. I was trying to get the shared memory matmul example working but for some reason kept getting incorrect results. Attached is 4 different examples of of the same matmul kernel and all should produce the same result. Only v4 produces the correct result for me, however.
[matmul.zip](https://github.com/numba/numba/files/4525103/matmul.zip)

**Comment by gmarkall** [comment]
The array `x` is typed as being in C-order. When it is typed as being in A-order, for example by eagerly compiling:

```diff
diff --git a/repro.py b/repro.py
index 174944b..fee416f 100644
--- a/repro.py
+++ b/repro.py
@@ -1,4 +1,4 @@
-from numba import cuda
+from numba import cuda, int64, void
 from numba.np import numpy_support as nps
 import numpy as np
 
@@ -7,7 +7,7 @@ nthreads = 16
 dt = nps.from_dtype(arr.dtype)
 nshared = nthreads * arr.dtype.itemsize
 
-@cuda.jit
+@cuda.jit(void(int64[:], int64))
 def sm_slice_copy(y, chunksize):
     dynsmem = cuda.shared.array(0, dtype=dt)
 
```

the issue disappears.

**Comment by gmarkall** [comment]
It turns out that [Emby](https://github.com/JonasRSV/emby) also contains code that can trigger a very similar issue to the original reproducer, and it is possible (likely) that the dynamic shared memory is not related to the source of the issue. Running this small example with Emby at commit https://github.com/JonasRSV/emby/commit/eb7659a7be2a576bf019b1a5dc9b4427f91b589b:

```python
import emby
from sklearn.datasets import load_digits

x = load_digits().data
som = emby.SOM(Z=2, bases=1000, device=emby.Device.GPU)
embeddings = som.fit_transform(x) 
```

with the following patch applied:

```diff
@@ -53,20 +53,20 @@ def _euclidean_argmin_gpu(arr_1: cuda.cudadrv.devicearray,
     grid_size = cuda.gridsize(1)
 
     
-    local = cuda.local.array(512000 // 32, dtype=float32) # max size 512kb
+    #local = cuda.local.array(512000 // 32, dtype=float32) # max size 512kb
     # using local memory gives a 2x speed-up
     # But it limits the dimension of the bases to 60 000
     # I don't believe this will be a problem though
     for i in range(arr_1_pos, arr_1_size, grid_size):
         min_distance, min_j = 1000000.0, 0
 
-        for k in range(dim): local[k] = arr_1[i, k] # pre-load
+        #for k in range(dim): local[k] = arr_1[i, k] # pre-load
 
         
         for j in range(arr_2_size):
             zj = 0.0
             for k in range(dim):
-                a1, a2 = local[k], arr_2[j, k]
+                a1, a2 = arr_1[i, k], arr_2[j, k]
                 zj += (a1 - a2) * (a1 - a2)
 
             if zj < min_distance:
@@ -97,7 +97,7 @@ def _euclidean_argmin(arr_1: np.ndarray, arr_2: np.ndarray):
     return res.copy_to_host()
 
 
-@cuda.jit
+@cuda.jit#(void(int64[:], float32[:,:], float32[:,:], float32[:,:], float32[:]))
 def _fit_winners_pull_gpu(winners: cuda.cudadrv.devicearray,
                           x: cuda.cudadrv.devicearray,
                           x_bases: cuda.cudadrv.devicearray,
@@ -111,18 +111,18 @@ def _fit_winners_pull_gpu(winners: cuda.cudadrv.devicearray,
 
     lr = learning_rate[0] # 20 % speedup by accesing that here
 
-    local = cuda.local.array(512000 // 64, dtype=float32) # 15% speed-up
+    #local = cuda.local.array(512000 // 64, dtype=float32) # 15% speed-up
     for i in range(winner_pos, winners_size, grid_size):
         winner = winners[i]
 
-        for k in range(dim): local[k] = x[i, k] # pre-load
+        #for k in range(dim): local[k] = x[i, k] # pre-load
 
         for j in range(bases_size):
             neighbourhood = y_neighbourhood[j, winner]
 
             for k in range(dim):
                 xb = x_bases[j, k]
-                x_bases[j, k] = xb + neighbourhood * lr * (local[k] - xb)
+                x_bases[j, k] = xb + neighbourhood * lr * (x[i, k] - xb)
 
 
 def _fit(x: np.ndarray,
```

(the local array and preloading is commented out because it seemed to cause some other issue) results in:

```
$ python gpu.py 
Traceback (most recent call last):
  File "gpu.py", line 6, in <module>
    embeddings = som.fit_transform(x) 
  File "/home/gmarkall/tmp/emby/emby/SOM/SOM.py", line 167, in fit_transform
    self.fit(x)
  File "/home/gmarkall/tmp/emby/emby/SOM/SOM.py", line 128, in fit
    verbose=self.fit_verbose)
  File "/home/gmarkall/tmp/emby/emby/SOM/gpu_core.py", line 168, in _fit
    cuda.synchronize()
  File "/home/gmarkall/miniconda3/envs/numbaenv/lib/python3.7/site-packages/numba-0.49.0-py3.7-linux-x86_64.egg/numba/cuda/api.py", line 198, in synchronize
    return current_context().synchronize()
  File "/home/gmarkall/miniconda3/envs/numbaenv/lib/python3.7/site-packages/numba-0.49.0-py3.7-linux-x86_64.egg/numba/cuda/cudadrv/driver.py", line 1162, in synchronize
    driver.cuCtxSynchronize()
  File "/home/gmarkall/miniconda3/envs/numbaenv/lib/python3.7/site-packages/numba-0.49.0-py3.7-linux-x86_64.egg/numba/cuda/cudadrv/driver.py", line 295, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/gmarkall/miniconda3/envs/numbaenv/lib/python3.7/site-packages/numba-0.49.0-py3.7-linux-x86_64.egg/numba/cuda/cudadrv/driver.py", line 330, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [700] Call to cuCtxSynchronize results in UNKNOWN_CUDA_ERROR
```

Running with cuda-memcheck gives:

```
$ cuda-gdb --args python gpu.py 
NVIDIA (R) CUDA Debugger
10.2 release
...
(cuda-gdb) set cuda memcheck on
(cuda-gdb) run
Starting program: /home/gmarkall/miniconda3/envs/numbaenv/bin/python gpu.py
...

Illegal access to address (@global)0x7ffeb5608000 detected.

Thread 1 "python" received signal CUDA_EXCEPTION_1, Lane Illegal Address.
[Switching focus to CUDA kernel 0, grid 3, block (0,0,0), thread (64,0,0), device 0, sm 0, warp 0, lane 0]
0x0000555557e183a0 in cudapy::emby::SOM::gpu_core::_fit_winners_pull_gpu$243(Array<long long, 1, C, mutable, aligned>, Array<float, 2, C, mutable, aligned>, Array<float, 2, C, mutable, aligned>, Array<float, 2, C, mutable, aligned>, Array<float, 1, C, mutable, aligned>) ()
(cuda-gdb) 
```

Uncommenting the signature on `_fit_winners_pull_gpu` (which forces the typing of arrays to be `A` instead of `C` layout) then enables running without error:

```
$ cuda-memcheck python gpu.py 
========= CUDA-MEMCHECK
========= ERROR SUMMARY: 0 errors
```

**Comment by gmarkall** [comment]
(cc @JonasRSV for visibility)

**Comment by gmarkall** [comment]
I'm thinking this might be to do with some mixing between signed and unsigned types, particularly to do with thread IDs and / or indices. This code:

```python
from numba import cuda
from numba.np import numpy_support as nps
import numpy as np

arr = np.arange(1024)
nthreads = 16
dt = nps.from_dtype(arr.dtype)
nshared = nthreads * arr.dtype.itemsize


@cuda.jit
def sm_slice_copy(y, chunksize):
    dynsmem = cuda.shared.array(0, dtype=dt)

    bx = cuda.blockIdx.x

    for j in range(chunksize):
        y[bx + j] = dynsmem[j]


sm_slice_copy[1, 1, 0, nshared](arr, 8)
```

produces:

```
Traceback (most recent call last):
  File "repro-for-issue.py", line 21, in <module>
    sm_slice_copy[1, 1, 0, nshared](arr, 8)
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 760, in __call__
    cfg(*args)
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 520, in __call__
    self._kernel_call(args=args,
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 635, in _kernel_call
    wb()
  File "/home/gmarkall/numbadev/numba/numba/cuda/args.py", line 65, in <lambda>
    retr.append(lambda: devary.copy_to_host(self.value, stream=stream))
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/devices.py", line 224, in _require_cuda_context
    return fn(*args, **kws)
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/devicearray.py", line 264, in copy_to_host
    _driver.device_to_host(hostary, self, self.alloc_size, stream=stream)
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 2269, in device_to_host
    fn(host_pointer(dst), device_pointer(src), size, *varargs)
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 295, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 330, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [700] Call to cuMemcpyDtoH results in UNKNOWN_CUDA_ERROR
```

and under memcheck:

```
========= CUDA-MEMCHECK
========= Invalid __global__ write of size 8
=========     at 0x00000950 in cudapy::__main__::sm_slice_copy$241(Array<__int64, int=1, C, mutable, aligned>, __int64)
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x7fd537800008 is out of bounds
=========     Device Frame:cudapy::__main__::sm_slice_copy$241(Array<__int64, int=1, C, mutable, aligned>, __int64) (cudapy::__main__::sm_slice_copy$241(Array<__int64, int=1, C, mutable, aligned>, __int64) : 0x950)
...
```

but when applying this patch to Numba:

```diff
diff --git a/numba/cuda/cudadecl.py b/numba/cuda/cudadecl.py
index dce7f429e..2c36e8884 100644
--- a/numba/cuda/cudadecl.py
+++ b/numba/cuda/cudadecl.py
@@ -21,12 +21,12 @@ class GridFunction(CallableTemplate):
         def typer(ndim):
             val = ndim.literal_value
             if val == 1:
-                restype = types.int32
+                restype = types.uint32
             elif val in (2, 3):
-                restype = types.UniTuple(types.int32, val)
+                restype = types.UniTuple(types.uint32, val)
             else:
                 raise ValueError('argument can only be 1, 2, 3')
-            return signature(restype, types.int32)
+            return signature(restype, types.uint32)
         return typer
 
 
@@ -331,13 +331,13 @@ class Dim3_attrs(AttributeTemplate):
     key = dim3
 
     def resolve_x(self, mod):
-        return types.int32
+        return types.uint32
 
     def resolve_y(self, mod):
-        return types.int32
+        return types.uint32
 
     def resolve_z(self, mod):
-        return types.int32
+        return types.uint32
 
 
 @register_attr
@@ -404,10 +404,10 @@ class CudaModuleTemplate(AttributeTemplate):
         return dim3
 
     def resolve_warpsize(self, mod):
-        return types.int32
+        return types.uint32
 
     def resolve_laneid(self, mod):
-        return types.int32
+        return types.uint32
 
     def resolve_shared(self, mod):
         return types.Module(cuda.shared)
```

it runs without error both in a normal run and under cuda-memcheck.

However that patch causes the reduction tests to fail, and does not fix `test_issue_5073` - so if some inconsistency like this is the cause, then there is still some inconsistency, or it introduces a new inconsistency.

**Comment by gmarkall** [comment]
Another data point - disabling IR optimization prior to sending the IR to NVVM with:

```diff
diff --git a/numba/cuda/codegen.py b/numba/cuda/codegen.py
index e201a2101..b0636be08 100644
--- a/numba/cuda/codegen.py
+++ b/numba/cuda/codegen.py
@@ -16,6 +16,7 @@ class CUDACodeLibrary(CodeLibrary):
         pass
 
     def _optimize_final_module(self):
+        return
         # Run some lightweight optimization to simplify the module.
         # This seems to workaround a libnvvm compilation bug (see #1341)
         pmb = ll.PassManagerBuilder()
```

seems to fix the issue - perhaps there is some slight semantic difference between LLVM versions that is introducing the problem when optimizing with a late version (9) and feeding to an earlier version (3.4-ish in NVVM).

**Comment by gmarkall** [comment]
PR to fix this by not optimizing with LLVM 9 prior to sending IR to NVVM is in #6030.

**Comment by gmarkall** [comment]
Now that #6030 is merged, this should in general be resolved, but there are a couple of things to check:

* [ ] Whether the [Emby code above](https://github.com/numba/numba/issues/5576#issuecomment-622427378) now works
* [ ] Whether applying the [patch to make thread / block ids unsigned](https://github.com/numba/numba/issues/5576#issuecomment-632808456) results in any errors.

**Comment by gmarkall** [comment]
Notes from investigations:

1. The Emby code above does now work on an RTX8000, and doesn't require the local array declarations to be modified either 
 - it's quite possible this was also something affected by the issues resolved in #6030. (cc @JonasRSV).
2. The patch above to make thread / block IDs doesn't automatically work now that #6030 is merged, but the issue was unrelated - there is an implicit expectation of warp synchronization at certain points in the reduction kernel (https://github.com/numba/numba/blob/master/numba/cuda/kernels/reduction.py#L30 and https://github.com/numba/numba/blob/master/numba/cuda/kernels/reduction.py#L37), which does not always hold.

Given the above, I believe this issue is now resolved and will close it.
